allennlp/data/dataset_readers/penn_tree_bank.py

```diff
@@ -0,0 +1,173 @@
+        tokens : ``List[str]``, required.
+            The tokens in a given sentence.
+        pos_tags ``List[str]``, required.
+            The pos tags for the words in the sentence.
+        gold_tree : ``Tree``, optional (default = None).
+            The gold parse tree to create span labels from.
+
+        Returns
+        -------
+        An ``Instance`` containing the following fields:
+            tokens : ``TextField``
+                The tokens in the sentence.
+            pos_tags : ``SequenceLabelField``
+                The pos tags of the words in the sentence.
+            spans : ``ListField[SpanField]``
+                A ListField containing all possible subspans of the
+                sentence.
+            span_labels : ``SequenceLabelField``, optional.
+                The constiutency tags for each of the possible spans, with
+                respect to a gold parse tree. If a span is not contained
+                within the tree, a span will have a ``NO-LABEL`` label.
+        """
+        # pylint: disable=arguments-differ
+        text_field = TextField([Token(x) for x in tokens], token_indexers=self._token_indexers)
+        pos_tag_field = SequenceLabelField(pos_tags, text_field, "pos_tags")
+        fields = {"tokens": text_field, "pos_tags": pos_tag_field}
+        spans: List[Field] = []
+        gold_labels = []
+
+        if gold_tree is not None:
+            gold_spans_with_pos_tags = self._get_gold_spans(gold_tree, 0, OrderedDict()).items()
+            gold_spans = {span: label for (span, label)
+                          in gold_spans_with_pos_tags if "-POS" not in label}
+        else:
+            gold_spans = None
+        for start, end in enumerate_spans(tokens):
+            spans.append(SpanField(start, end, text_field))
+
+            if gold_spans is not None:
+                if (start, end) in gold_spans.keys():
+                    gold_labels.append(gold_spans[(start, end)])
+                else:
+                    gold_labels.append("NO-LABEL")
+
+        span_list_field: ListField = ListField(spans)
+        fields["spans"] = span_list_field
+        if gold_tree is not None:
+            fields["span_labels"] = SequenceLabelField(gold_labels, span_list_field)
+
+        return Instance(fields)
+
+    def _get_gold_spans(self, # pylint: disable=arguments-differ
+                        tree: Tree,
+                        index: int,
+                        typed_spans: OrderedDict) -> OrderedDict:
+        """
+        Recursively construct the gold spans from an nltk ``Tree``.
+        Spans are inclusive.
+
+        Parameters
+        ----------
+        tree : ``Tree``, required.
+            An NLTK parse tree to extract spans from.
+        index : ``int``, required.
+            The index of the current span in the sentence being considered.
+        typed_spans : ``OrderedDict[Tuple[int, int], str]``, required.
+            A ordered dictionary mapping spans to span labels.
```

Comment:
You might give a note here that you're using an `OrderedDict` because you rely on its efficiency in the method implementation.  I had to read the comments to know why this wasn't just a `Dict`.

Reply:
It's mostly because of that, but python `Dict`s don't implement `__reversed__()`, so even though they have a deterministic order, you have to do something like `next(reversed(list(dict.keys())))`. But point taken, i'll add a comment.